src/Native/System.Security.Cryptography.Native/pal_ssl.cpp

```diff
@@ -303,209 +304,158 @@ extern "C" int32_t CryptoNative_SslSessionReused(SSL* ssl)
 
-/*
-The values used in OpenSSL for SSL_CIPHER algorithm_mkey.
-*/
-enum class SSL_KeyExchangeAlgorithm : int64_t
-{
-    SSL_kRSA = 1,
-    /* DH cert, RSA CA cert */
-    SSL_kDHr = 2,
-    /* DH cert, DSA CA cert */
-    SSL_kDHd = 4,
-    /* tmp DH key no DH cert */
-    SSL_kEDH = 8,
-    /* Kerberos5 key exchange */
-    SSL_kKRB5 = 16,
-    /* ECDH cert, RSA CA cert */
-    SSL_kECDHr = 32,
-    /* ECDH cert, ECDSA CA cert */
-    SSL_kECDHe = 64,
-    SSL_kEECDH = 128,
-    SSL_kPSK = 256,
-    SSL_kGOST = 512,
-    SSL_kSRP = 1024,
-};
-
-static ExchangeAlgorithmType MapExchangeAlgorithmType(const SSL_CIPHER* cipher)
-{
-    unsigned long mkey = cipher->algorithm_mkey;
-
-    SSL_KeyExchangeAlgorithm sslMkey = static_cast<SSL_KeyExchangeAlgorithm>(mkey);
-    switch (sslMkey)
-    {
-        case SSL_KeyExchangeAlgorithm::SSL_kRSA:
-            return ExchangeAlgorithmType::RsaKeyX;
-
-        case SSL_KeyExchangeAlgorithm::SSL_kDHr:
-            return ExchangeAlgorithmType::DiffieHellman;
-
-        case SSL_KeyExchangeAlgorithm::SSL_kDHd:
-            return ExchangeAlgorithmType::DiffieHellman;
-
-        case SSL_KeyExchangeAlgorithm::SSL_kEDH:
-            return ExchangeAlgorithmType::DiffieHellman;
+static ExchangeAlgorithmType MapExchangeAlgorithmType(const std::string& keyExchange)
+{
+    if (keyExchange.compare("RSA") == 0)
+        return ExchangeAlgorithmType::RsaKeyX;
+    if (keyExchange.compare("DH/RSA") == 0)
+        return ExchangeAlgorithmType::DiffieHellman;
+    if (keyExchange.compare("DH/DSS") == 0)
+        return ExchangeAlgorithmType::DiffieHellman;
+    if (keyExchange.compare("DH") == 0)
+        return ExchangeAlgorithmType::DiffieHellman;
+    if (keyExchange.compare("KRB5") == 0)
+       return ExchangeAlgorithmType::SSL_kKRB5;
+    if (keyExchange.compare("ECDH/RSA") == 0)
+        return ExchangeAlgorithmType::SSL_ECDH;
+    if (keyExchange.compare("ECDH/ECDSA") == 0)
+        return ExchangeAlgorithmType::SSL_ECDSA;
+    if (keyExchange.compare("ECDH") == 0)
+       return ExchangeAlgorithmType::SSL_ECDSA;
+    if (keyExchange.compare("PSK") == 0)
+        return ExchangeAlgorithmType::SSL_kPSK;
+    if (keyExchange.compare("GOST") == 0)
+        return ExchangeAlgorithmType::SSL_kGOST;
+    if (keyExchange.compare("SRP") == 0)
+        return ExchangeAlgorithmType::SSL_kSRP;
 
-        case SSL_KeyExchangeAlgorithm::SSL_kKRB5:
-            return ExchangeAlgorithmType::SSL_kKRB5;
-
-        case SSL_KeyExchangeAlgorithm::SSL_kECDHr:
-            return ExchangeAlgorithmType::SSL_ECDH;
-
-        case SSL_KeyExchangeAlgorithm::SSL_kECDHe:
-            return ExchangeAlgorithmType::SSL_ECDSA;
+    return ExchangeAlgorithmType::None;
+}
 
-        case SSL_KeyExchangeAlgorithm::SSL_kEECDH:
-            return ExchangeAlgorithmType::SSL_ECDSA;
+static void GetHashAlgorithmTypeAndSize(const std::string& mac, HashAlgorithmType* dataHashAlg, DataHashSize* hashKeySize)
+{
+    if (mac.compare("MD5") == 0)
```

Comment:
There are already checks for the pointers in the extern method:

https://github.com/dotnet/corefx/blob/master/src/Native/System.Security.Cryptography.Native/pal_ssl.cpp#L376-L379